<a href="https://colab.research.google.com/github/AndyMG1999/Csc448Andy/blob/main/Assignment_4_Andy_Gonzalez.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [40]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.linear_model import BayesianRidge
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures as poly
from sklearn.model_selection import ShuffleSplit
from sklearn.datasets import load_iris
from sklearn import svm
from sklearn.svm import SVC
from sklearn.neural_network import MLPClassifier
from sklearn.neighbors import KNeighborsClassifier

# Part 1: Load the dataset

In [48]:
# Load the dataset (load remotely, not locally)
iris = load_iris()
iris_df = pd.DataFrame(iris['data'],columns = iris.feature_names)
iris_df["species"]= iris.target

# Output the first 15 rows of the data
iris_df.rename(columns = {'Sepal Length':'0'})
iris_df.head(15)

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                 5.1               3.5                1.4               0.2   
1                 4.9               3.0                1.4               0.2   
2                 4.7               3.2                1.3               0.2   
3                 4.6               3.1                1.5               0.2   
4                 5.0               3.6                1.4               0.2   
5                 5.4               3.9                1.7               0.4   
6                 4.6               3.4                1.4               0.3   
7                 5.0               3.4                1.5               0.2   
8                 4.4               2.9                1.4               0.2   
9                 4.9               3.1                1.5               0.1   
10                5.4               3.7                1.5               0.2   
11                4.8               3.4                1.6               0.2   
12                4.8               3.0                1.4               0.1   
13                4.3               3.0                1.1               0.1   
14                5.8               4.0                1.2               0.2   

    species  
0         0  
1         0  
2         0  
3         0  
4         0  
5         0  
6         0  
7         0  
8         0  
9         0  
10        0  
11        0  
12        0  
13        0  
14        0

In [42]:
# Display a summary of the table information (number of datapoints, etc.)
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
 4   species            150 non-null    int64  
dtypes: float64(4), int64(1)
memory usage: 6.0 KB


## About the dataset
Explain what the data is in your own words. What are your features and labels? What is the mapping of your labels to the actual classes?

The data containers different measurements of iris plant's sepal and petal sizes. The features are these sizes and I am assuming the labels are the species of iris it would be. I assume the ML programs are going to find patterns in this data and make predictions as to what the species would be based on the data.

# Part 2: Split the dataset into train and test

In [49]:
# Take the dataset and split it into our features (X) and label (y)
features = iris_df[["sepal length (cm)","sepal width (cm)","petal length (cm)","petal width (cm)"]]
label = iris_df['species']
x, y = features, label

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=.10)
x_train, x_test, y_train, y_test

(     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
 18                 5.7               3.8                1.7               0.3
 100                6.3               3.3                6.0               2.5
 94                 5.6               2.7                4.2               1.3
 3                  4.6               3.1                1.5               0.2
 24                 4.8               3.4                1.9               0.2
 ..                 ...               ...                ...               ...
 102                7.1               3.0                5.9               2.1
 118                7.7               2.6                6.9               2.3
 98                 5.1               2.5                3.0               1.1
 134                6.1               2.6                5.6               1.4
 78                 6.0               2.9                4.5               1.5
 
 [135 rows x 4 columns],
      sepal length (cm)  

# Part 3: Logistic Regression

In [44]:
# i. Use sklearn to train a LogisticRegression model on the training set
log_reg = LogisticRegression(max_iter=1000)
log_reg.fit(x_train,y_train)

LogisticRegression(max_iter=1000)

In [50]:
# ii. For a sample datapoint, predict the probabilities for each possible class
log_reg.predict(np.array([[1,2,3,4]]))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  "X does not have valid feature names, but"


array([2])

2 represents the iris species Virgincia 

In [51]:
# ii. For a sample datapoint, predict the probabilities for each possible class
log_reg.predict_proba(np.array([[1,2,3,4]]))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  "X does not have valid feature names, but"


array([[0.04805288, 0.00867178, 0.94327535]])

In [52]:
# iii. Report on the score for Logistic regression model, what does the score measure?
log_reg.score(x_test,y_test)

0.9333333333333333

The score measures our chances(percentage) of classifying the flower correctly.

In [53]:
# iv. Extract the coefficents and intercepts for the boundary line(s)
print("Coeficients: ",log_reg.coef_)
print("Intercept: ",log_reg.intercept_)

Coeficients:  [[-0.4360759   0.91939637 -2.44774077 -1.05167985]
 [ 0.47531899 -0.27959438 -0.14858175 -0.94978785]
 [-0.03924309 -0.63980198  2.59632252  2.0014677 ]]
Intercept:  [  9.72561566   2.19492867 -11.92054434]


# Part 4: Support Vector Machine

In [54]:
# i. Use sklearn to train a Support Vector Classifier on the training set
svm_model = SVC(probability=True)
SVC.fit(svm_model,x_train,y_train)

SVC(probability=True)

In [55]:
# ii. For a sample datapoint, predict the probabilities for each possible class
svm_model.predict([[1,2,3,4]])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  "X does not have valid feature names, but"


array([2])

In [56]:
# ii. For a sample datapoint, predict the probabilities for each possible class
svm_model.predict_proba([[1,2,3,4]])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but SVC was fitted with feature names
  "X does not have valid feature names, but"


array([[0.29902652, 0.32038906, 0.38058442]])

In [57]:
# iii. Report on the score for the SVM, what does the score measure?
svm_model.score(x_test,y_test)

0.9333333333333333

The score measures our chances(percentage) of classifying the flower correctly.

# Part 5: Neural Network

In [59]:
# i. Use sklearn to train a Neural Network (MLP Classifier) on the training set
mlp_model = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes =(15,), random_state=1).fit(features,label)
mlp_model.fit(x_train,y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
mlp_model.predict_proba([[1,2,3,4]])


/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but MLPClassifier was fitted with feature names
  "X does not have valid feature names, but"


array([[0.00000000e+00, 2.47997765e-13, 1.00000000e+00]])

In [60]:
# iii. Report on the score for the Neural Network, what does the score measure?
mlp_model.score(x_test,y_test)
# iv: Experiment with different options for the neural network, report on your best configuration (the highest score I was able to achieve was 0.8666)

1.0

The score measures our chances(percentage) of classifying the flower correctly.

# Part 6: K-Nearest Neighbors

In [61]:
# i. Use sklearn to 'train' a k-Neighbors Classifier
# Note: KNN is a nonparametric model and technically doesn't require training
# fit will essentially load the data into the model see link below for more information
# https://stats.stackexchange.com/questions/349842/why-do-we-need-to-fit-a-k-nearest-neighbors-classifier
kNN_model = KNeighborsClassifier(n_neighbors=5)
kNN_model.fit(x_train,y_train)
# ii. For a sample datapoint, predict the probabilities for each possible class
kNN_model.predict_proba([[1,2,3,4]])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but KNeighborsClassifier was fitted with feature names
  "X does not have valid feature names, but"


array([[0., 1., 0.]])

In [62]:
# iii. Report on the score for kNN, what does the score measure?
kNN_model.score(x_test,y_test)

0.9333333333333333

# Part 7: Conclusions and takeaways

In your own words describe the results of the notebook. Which model(s) performed the best on the dataset? Why do you think that is? Did anything surprise you about the exercise?


The main goal of the notebook was to accurately classify different species of iris plants. The dataset that was used contained different measurements for iris sepal length/width and petal length/width in order for the the ML to take this data and find patterns between the data. Out of all the models, the neural network model worked best with an accuracy of 100% while every other model scored a 93%. If these scores are accurate I can see why deep learning has been the trending topic within the AI space lately. Neural networks are known for excelling at pattern recognition and allow the creation of very complex functions to fit the data. I think what surprised me most about this excercise is how available these very powerful tools are available to everyone. Tools like these tend to feel almost like legends, tools that would be mentioned constantly in lectures but only used by NASA engineers or Ivy League Scientists. It was nice to see what we went over during lectures in practice.